assets/app/view/game/part/location_name.rb

```diff
@@ -42,7 +46,29 @@ def preferred_render_locations
               end
             end
 
-            return [center, l_up40, l_down40] if @tile.exits.empty? && @tile.cities.empty?
+            # flat map: if two cities, and they are on edges 0 and 3, pick the center.
+            if layout == :flat && @tile.cities.size == 2 &&
+                ([0, 3] - @tile.city_towns.map { |c| @tile.preferred_city_town_edges[c] }).empty?
+              return [center]
+            end
+
+            # pointy map: if two cities or towns, and neither of them are on edges 1 or 4 or in the center,
+            # pick the center.
+            if layout == :pointy && @tile.city_towns.size == 2 &&
+                ([1, 4] - @tile.city_towns.map { |ct| @tile.preferred_city_town_edges[ct] }).size == 2 &&
```

Comment:
should we store @tile.city-towns.map above since you call it here and on line 51

Reply:
OK